operator/internal/securedcluster/extensions/rollout_restart_sensor_ca.go

```diff
@@ -0,0 +1,205 @@
+package extensions
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/operator-framework/helm-operator-plugins/pkg/extensions"
+	"github.com/pkg/errors"
+	commonAnnotations "github.com/stackrox/rox/operator/internal/common/annotations"
+	"github.com/stackrox/rox/operator/internal/utils"
+	"github.com/stackrox/rox/pkg/crs"
+	"github.com/stackrox/rox/pkg/securedcluster"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var allTLSSecretNames = []string{
+	securedcluster.CollectorTLSSecretName,
+	securedcluster.AdmissionControlTLSSecretName,
+	securedcluster.ScannerTLSSecretName,
+	securedcluster.ScannerDbTLSSecretName,
+	securedcluster.ScannerV4IndexerTLSSecretName,
+	securedcluster.ScannerV4DbTLSSecretName,
+}
+
+// RolloutRestartOnSensorCAChange is an extension that triggers rollout restarts of workloads
+// when the Sensor CA changes.
+func RolloutRestartOnSensorCAChange(client ctrlClient.Client, direct ctrlClient.Reader, logger logr.Logger) extensions.ReconcileExtension {
+	return func(ctx context.Context, obj *unstructured.Unstructured, statusUpdater func(statusFunc extensions.UpdateStatusFunc), log logr.Logger) error {
+		logger = logger.WithName("rollout-restart-sensor-ca")
+
+		sensorHash, fromRuntimeSecret, err := tryGetSensorCAHash(ctx, client, direct, obj.GetNamespace())
+		if err != nil {
+			return err
+		}
+
+		annotations := getOrInitAnnotations(obj)
+
+		// runtime TLS secrets are not stored atomically, so we should wait for consistency
+		if fromRuntimeSecret {
+			if err := waitForCAConsistency(ctx, client, direct, obj.GetNamespace(), sensorHash, logger); err != nil {
```

Comment:
Blocking in the reconciler is frowned upon because it reduces the controller's throughput..
Also, is proceeding here despite inconsistent CA a good idea? Won't changing the annotation in this case basically guarantee that the components with stale secrets get in trouble? 🤔

Reply:
I dropped the blocking part, and failing the reconcile if the certificates are not consistent. Not sure if that's the best approach, or simply ignoring would be better.